File: zero-server/src/main/kotlin/com/zeroday/model/Research.kt
package com.zeroday.model

import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class InventoryItem(
    val id: String = UUID.randomUUID().toString(),
    val type: ItemType,
    val name: String,
    val description: String,
    val quantity: Int = 1,
    val rarity: ItemRarity = ItemRarity.COMMON,
    val data: Map<String, String> = emptyMap()
)

@Serializable
enum class ItemType {
    KNOWLEDGE_FRAGMENT,
    ZERO_DAY_EXPLOIT,
    CRYPT_KEY,
    ACCESS_TOKEN,
    MALWARE_PAYLOAD,
    DECOY_DATA,
    FIREWALL_PATCH
}

@Serializable
enum class ItemRarity(val displayName: String, val color: String, val multiplier: Double) {
    COMMON("Common", "#808080", 1.0),
    UNCOMMON("Uncommon", "#00FF00", 1.5),
    RARE("Rare", "#4488FF", 2.5),
    EPIC("Epic", "#AA44FF", 4.0),
    LEGENDARY("Legendary", "#FF8800", 8.0)
}

fun InventoryItem.toMap(): Map<String, Any?> = mapOf(
    "id" to id,
    "type" to type.name,
    "name" to name,
    "description" to description,
    "quantity" to quantity,
    "rarity" to rarity.name,
    "data" to data
)

@Serializable
data class CraftingRecipe(
    val id: String,
    val name: String,
    val description: String,
    val resultType: ItemType,
    val resultName: String,
    val resultRarity: ItemRarity,
    val ingredients: List<Ingredient>,
    val craftingTimeMs: Long = 10_000L,
    val requiredLevel: Int = 1,
    val requiredCommands: List<String> = emptyList(),
    val xpReward: Long = 100,
    val resultDescription: String = ""
)

@Serializable
data class Ingredient(
    val itemType: ItemType,
    val quantity: Int = 1,
    val minRarity: ItemRarity = ItemRarity.COMMON,
    val creditsCost: Long = 0
)

@Serializable
data class ResearchProgress(
    val recipeId: String,
    val startedAt: Long = System.currentTimeMillis(),
    var completed: Boolean = false,
    var claimed: Boolean = false
)

object CraftingRecipes {
    val allRecipes: List<CraftingRecipe> = listOf(
        CraftingRecipe(
            "fragment_to_crypt", "Fragment Compression",
            "Compress raw knowledge fragments into a reusable cryptographic key",
            ItemType.CRYPT_KEY, "Encrypted Cipher Key", ItemRarity.UNCOMMON,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 3)),
            8000L, 2, xpReward = 50,
            resultDescription = "A reusable key that boosts decrypt command strength"
        ),
        CraftingRecipe(
            "fragment_to_token", "Access Token Fabrication",
            "Forge an access token from network intelligence fragments",
            ItemType.ACCESS_TOKEN, "Skeleton Admin Token", ItemRarity.RARE,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 5), Ingredient(ItemType.CRYPT_KEY, 1, ItemRarity.UNCOMMON)),
            15000L, 4, listOf("decrypt"), xpReward = 150,
            resultDescription = "Grants temporary admin access to any discovered node"
        ),
        CraftingRecipe(
            "zero_day_basic", "Basic Zero-Day Exploit",
            "Craft a single-use zero-day exploit from vulnerability research",
            ItemType.ZERO_DAY_EXPLOIT, "CVE-2024-Local", ItemRarity.RARE,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 8), Ingredient(ItemType.ACCESS_TOKEN, 1, ItemRarity.UNCOMMON)),
            30000L, 6, listOf("exploit"), xpReward = 300,
            resultDescription = "Bypasses security level ≤5 on any target (single use)"
        ),
        CraftingRecipe(
            "zero_day_advanced", "Advanced Zero-Day Exploit",
            "Combine rare fragments and crypt keys for a powerful exploit",
            ItemType.ZERO_DAY_EXPLOIT, "CVE-2024-Remote-Code-Execution", ItemRarity.EPIC,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 15), Ingredient(ItemType.CRYPT_KEY, 3, ItemRarity.RARE),
                   Ingredient(ItemType.MALWARE_PAYLOAD, 1, ItemRarity.UNCOMMON)),
            60000L, 10, listOf("exploit", "backdoor"), xpReward = 800,
            resultDescription = "Bypasses security level ≤10 on any target (single use)"
        ),
        CraftingRecipe(
            "zero_day_legendary", "Legendary Zero-Day Exploit",
            "The ultimate weapon - bypasses any security, leaves no trace",
            ItemType.ZERO_DAY_EXPLOIT, "CVE-2024-God-Mode", ItemRarity.LEGENDARY,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 50), Ingredient(ItemType.CRYPT_KEY, 5, ItemRarity.EPIC),
                   Ingredient(ItemType.ZERO_DAY_EXPLOIT, 2, ItemRarity.EPIC), Ingredient(ItemType.ACCESS_TOKEN, 3, ItemRarity.EPIC)),
            180000L, 16, listOf("exploit", "backdoor", "rootkit"), xpReward = 2500,
            resultDescription = "Bypasses ALL security. Permanent access granted. No trace."
        ),
        CraftingRecipe(
            "malware_payload", "Custom Malware Payload",
            "Engineer a custom malware payload from research data",
            ItemType.MALWARE_PAYLOAD, "RAT v3.0 Custom", ItemRarity.UNCOMMON,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 4)),
            12000L, 3, listOf("backdoor"), xpReward = 100,
            resultDescription = "A reusable payload for backdoor and worm commands"
        ),
        CraftingRecipe(
            "firewall_patch", "Firewall Hardening Patch",
            "Create a defensive patch that boosts firewall command effectiveness",
            ItemType.FIREWALL_PATCH, "ICE-Breaker Patch", ItemRarity.RARE,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 6), Ingredient(ItemType.DECOY_DATA, 1, ItemRarity.UNCOMMON)),
            20000L, 5, listOf("firewall"), xpReward = 200,
            resultDescription = "Blocks 3 additional trace attempts on your system"
        ),
        CraftingRecipe(
            "decoy_data", "Generate Decoy Data",
            "Create convincing decoy files to mislead追踪ers",
            ItemType.DECOY_DATA, "Fake Credentials Bundle", ItemRarity.COMMON,
            listOf(Ingredient(ItemType.KNOWLEDGE_FRAGMENT, 2)),
            5000L, 1, xpReward = 25,
            resultDescription = "Deploy decoy to waste attacker time and resources"
        )
    )

    fun getAvailableRecipes(level: Int, unlockedCommands: Set<String>): List<CraftingRecipe> =
        allRecipes.filter { recipe ->
            recipe.requiredLevel <= level &&
            recipe.requiredCommands.all { it in unlockedCommands }
        }
}

File: zero-server-csharp/src/ZeroDayMMO.Shared/Models/Research.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ZeroDayMMO.Shared.Models;

public class CraftingRecipe
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("result_type")]
    public ItemType ResultType { get; set; }

    [JsonPropertyName("result_name")]
    public string ResultName { get; set; } = "";

    [JsonPropertyName("result_rarity")]
    public ItemRarity ResultRarity { get; set; }

    [JsonPropertyName("ingredients")]
    public List<Ingredient> Ingredients { get; set; } = new();

    [JsonPropertyName("crafting_time_ms")]
    public long CraftingTimeMs { get; set; } = 10000;

    [JsonPropertyName("required_level")]
    public int RequiredLevel { get; set; } = 1;

    [JsonPropertyName("required_commands")]
    public List<string> RequiredCommands { get; set; } = new();

    [JsonPropertyName("xp_reward")]
    public long XpReward { get; set; } = 100;

    [JsonPropertyName("result_description")]
    public string ResultDescription { get; set; } = "";
}

public class Ingredient
{
    [JsonPropertyName("item_type")]
    public ItemType ItemType { get; set; }

    [JsonPropertyName("quantity")]
    public int Quantity { get; set; } = 1;

    [JsonPropertyName("min_rarity")]
    public ItemRarity MinRarity { get; set; } = ItemRarity.COMMON;

    [JsonPropertyName("credits_cost")]
    public long CreditsCost { get; set; } = 0;
}

public class ResearchProgress
{
    [JsonPropertyName("recipe_id")]
    public string RecipeId { get; set; } = "";

    [JsonPropertyName("started_at")]
    public long StartedAt { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    [JsonPropertyName("completed")]
    public bool Completed { get; set; } = false;

    [JsonPropertyName("claimed")]
    public bool Claimed { get; set; } = false;
}

public static class CraftingRecipes
{
    public static List<CraftingRecipe> AllRecipes { get; } = new()
    {
        new CraftingRecipe
        {
            Id = "fragment_to_crypt", Name = "Fragment Compression",
            Description = "Compress raw knowledge fragments into a reusable cryptographic key",
            ResultType = ItemType.CRYPT_KEY, ResultName = "Encrypted Cipher Key", ResultRarity = ItemRarity.UNCOMMON,
            Ingredients = new() { new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 3 } },
            CraftingTimeMs = 8000, RequiredLevel = 2, XpReward = 50,
            ResultDescription = "A reusable key that boosts decrypt command strength"
        },
        new CraftingRecipe
        {
            Id = "fragment_to_token", Name = "Access Token Fabrication",
            Description = "Forge an access token from network intelligence fragments",
            ResultType = ItemType.ACCESS_TOKEN, ResultName = "Skeleton Admin Token", ResultRarity = ItemRarity.RARE,
            Ingredients = new()
            {
                new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 5 },
                new Ingredient { ItemType = ItemType.CRYPT_KEY, Quantity = 1, MinRarity = ItemRarity.UNCOMMON }
            },
            CraftingTimeMs = 15000, RequiredLevel = 4, RequiredCommands = new() { "decrypt" }, XpReward = 150,
            ResultDescription = "Grants temporary admin access to any discovered node"
        },
        new CraftingRecipe
        {
            Id = "zero_day_basic", Name = "Basic Zero-Day Exploit",
            Description = "Craft a single-use zero-day exploit from vulnerability research",
            ResultType = ItemType.ZERO_DAY_EXPLOIT, ResultName = "CVE-2024-Local", ResultRarity = ItemRarity.RARE,
            Ingredients = new()
            {
                new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 8 },
                new Ingredient { ItemType = ItemType.ACCESS_TOKEN, Quantity = 1, MinRarity = ItemRarity.UNCOMMON }
            },
            CraftingTimeMs = 30000, RequiredLevel = 6, RequiredCommands = new() { "exploit" }, XpReward = 300,
            ResultDescription = "Bypasses security level ≤5 on any target (single use)"
        },
        new CraftingRecipe
        {
            Id = "zero_day_advanced", Name = "Advanced Zero-Day Exploit",
            Description = "Combine rare fragments and crypt keys for a powerful exploit",
            ResultType = ItemType.ZERO_DAY_EXPLOIT, ResultName = "CVE-2024-Remote-Code-Execution", ResultRarity = ItemRarity.EPIC,
            Ingredients = new()
            {
                new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 15 },
                new Ingredient { ItemType = ItemType.CRYPT_KEY, Quantity = 3, MinRarity = ItemRarity.RARE },
                new Ingredient { ItemType = ItemType.MALWARE_PAYLOAD, Quantity = 1, MinRarity = ItemRarity.UNCOMMON }
            },
            CraftingTimeMs = 60000, RequiredLevel = 10, RequiredCommands = new() { "exploit", "backdoor" }, XpReward = 800,
            ResultDescription = "Bypasses security level ≤10 on any target (single use)"
        },
        new CraftingRecipe
        {
            Id = "zero_day_legendary", Name = "Legendary Zero-Day Exploit",
            Description = "The ultimate weapon - bypasses any security, leaves no trace",
            ResultType = ItemType.ZERO_DAY_EXPLOIT, ResultName = "CVE-2024-God-Mode", ResultRarity = ItemRarity.LEGENDARY,
            Ingredients = new()
            {
                new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 50 },
                new Ingredient { ItemType = ItemType.CRYPT_KEY, Quantity = 5, MinRarity = ItemRarity.EPIC },
                new Ingredient { ItemType = ItemType.ZERO_DAY_EXPLOIT, Quantity = 2, MinRarity = ItemRarity.EPIC },
                new Ingredient { ItemType = ItemType.ACCESS_TOKEN, Quantity = 3, MinRarity = ItemRarity.EPIC }
            },
            CraftingTimeMs = 180000, RequiredLevel = 16, RequiredCommands = new() { "exploit", "backdoor", "rootkit" }, XpReward = 2500,
            ResultDescription = "Bypasses ALL security. Permanent access granted. No trace."
        },
        new CraftingRecipe
        {
            Id = "malware_payload", Name = "Custom Malware Payload",
            Description = "Engineer a custom malware payload from research data",
            ResultType = ItemType.MALWARE_PAYLOAD, ResultName = "RAT v3.0 Custom", ResultRarity = ItemRarity.UNCOMMON,
            Ingredients = new() { new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 4 } },
            CraftingTimeMs = 12000, RequiredLevel = 3, RequiredCommands = new() { "backdoor" }, XpReward = 100,
            ResultDescription = "A reusable payload for backdoor and worm commands"
        },
        new CraftingRecipe
        {
            Id = "firewall_patch", Name = "Firewall Hardening Patch",
            Description = "Create a defensive patch that boosts firewall command effectiveness",
            ResultType = ItemType.FIREWALL_PATCH, ResultName = "ICE-Breaker Patch", ResultRarity = ItemRarity.RARE,
            Ingredients = new()
            {
                new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 6 },
                new Ingredient { ItemType = ItemType.DECOY_DATA, Quantity = 1, MinRarity = ItemRarity.UNCOMMON }
            },
            CraftingTimeMs = 20000, RequiredLevel = 5, RequiredCommands = new() { "firewall" }, XpReward = 200,
            ResultDescription = "Blocks 3 additional trace attempts on your system"
        },
        new CraftingRecipe
        {
            Id = "decoy_data", Name = "Generate Decoy Data",
            Description = "Create convincing decoy files to mislead追踪ers",
            ResultType = ItemType.DECOY_DATA, ResultName = "Fake Credentials Bundle", ResultRarity = ItemRarity.COMMON,
            Ingredients = new() { new Ingredient { ItemType = ItemType.KNOWLEDGE_FRAGMENT, Quantity = 2 } },
            CraftingTimeMs = 5000, RequiredLevel = 1, XpReward = 25,
            ResultDescription = "Deploy decoy to waste attacker time and resources"
        }
    };

    public static List<CraftingRecipe> GetAvailableRecipes(int level, HashSet<string> unlockedCommands) =>
        AllRecipes.Where(recipe =>
            recipe.RequiredLevel <= level &&
            recipe.RequiredCommands.All(c => unlockedCommands.Contains(c))
        ).ToList();
}
